Prevent re-creating session file on end

blueo opened the pull request:
`saveState` will re-create a session file after the test session has ended. This adds a condition to check if the test session has been cleared before writing the file.

blueo commented:
> feels like it might be better to perform that kind of checks outside this method so we keep it simple and its semantics reflect its name and concern

Thanks @dnsl48, yeah I wasn't totally happy with this. What do you think of splitting the `saveState` out from `applyState`. So making `applyState` just return the newly modified state which is then passed to `saveState` for persistence. Would add another call to everywhere `applyState` is currently used but would allow checks before saving and might just be a bit cleaner.

dnsl48 commented:
> What do you think of splitting the saveState out from applyState

Sounds like a plan. I guess we don't really have a choice right? :)

dnsl48 commented:
> making applyState just return the newly modified state 

it also feels weird that `applyState` actually modifies the state... Since you're looking into it anyway, could you please consider if we might refactor this too?

blueo commented:
> > making applyState just return the newly modified state
> 
> it also feels weird that `applyState` actually modifies the state... Since you're looking into it anyway, could you please consider if we might refactor this too?

I was thinking`applyState` will return a new copy of the state - not touching the 'original' state - is that what you mean?

dnsl48 commented:
> I was thinkingapplyState will return a new copy of the state

well, I was rather thinking about moving out of it any modifications performed on the state, so it would only connect to the database preliminarily created outside. So it would only "apply" and not actually "modify" $state anyhow

chillu commented:
@blueo Do you agree with Serge, and can implement his suggestion?

blueo commented:
Hey @chillu i was looking into it but haven't quite managed to get all state change out. the biggest hurdle being that the test db name may change. But in general yes I agree and I'll be working on re-factoring it the next chance I get :)

blueo commented:
Hi @dnsl48 :) So that turned into a bit of a rabbit hole. I've refactored to separate our `applyState` and `saveState`. In the process I've actually addressed the other two PRs I have open https://github.com/silverstripe/silverstripe-testsession/pull/64 and https://github.com/silverstripe/silverstripe-testsession/pull/63 (as they'd need significant change also).

given it is quite a big change I've added some unit tests to try cover some basic scenarios - i don't think this is exaustive atm. I'm using this refactor on an internal project so will be able to beef up the tests a bit as I go. cc @chillu

dnsl48 commented:
wow, that definitely looks cool! I'll have a play with it some time this week.

dnsl48 commented:
Hey @blueo, sorry for delay. I've run it, but it breaks otherwise green tests on my local setup.
That's what I'm getting:
```
./vendor/bin/behat @framework                                                                                                             
Bootstrapping
Creating test session environment                                                                                                                                             
Temp Database: ss_tmpdb_1557720566_5399227
                                                                                       
@retry                                                                                 
Feature: Log in
  As an site owner               
  I want to access to the CMS to be secure
  So that only my team can make content changes
                                           
  Scenario: Bad login                                                 # vendor/silverstripe/admin/tests/behat/features/login.feature:7
    Given I log in with "bad@example.com" and "badpassword"           # SilverStripe\BehatExtension\Context\LoginContext::stepILogInWith()
    │                                     
    ╳  Notice: Trying to get property of non-object in vendor/silverstripe/testsession/src/TestSessionEnvironment.php line 648
    │
    └─ @AfterStep # SilverStripe\Framework\Tests\Behaviour\FeatureContext::waitResponsesAfterStep()                                   
    Then I should see "The provided details don't seem to be correct" # SilverStripe\Framework\Tests\Behaviour\FeatureContext::assertPageContainsText()
```

I'll do some debugging later this week, but it seems like `TestSessionState` isn't being persisted properly anymore.

blueo commented:
ah thanks @dnsl48 I'll have a look, i should probably add some basic behat checks

blueo commented:
Hey there, @dnsl48 yep, I missed creating the testsession. I also had a broken mailer setup, I've addressed both and added some test cases for it 👍 framework behat now runs green for me

dnsl48 commented:
Nice! Do you mind resolving the conflict or rebasing the branch yet?

blueo commented:
> Nice! Do you mind resolving the conflict or rebasing the branch yet?

done :)

dnsl48 commented on tests/unit/TestSessionControllerTest.php:
```diff
@@ -0,0 +1,278 @@
+
+    protected $usesDatabase = true;
+
+    protected function setUp(): void
```
yeah, I'm keen on that too, but we still need to support PHP 5.6 for some while.

```php
Parse error: syntax error, unexpected ':', expecting ';' or '{' in /home/vagrant/bench/ins44/vendor/silverstripe/testsession/tests/unit/TestSessionControllerTest.php on line 31 
```

dnsl48 commented:
I know that's not changed in the PR, but could we change it to
`!is_subclass_of($mailer, Mailer::class)`?
https://github.com/silverstripe/silverstripe-testsession/blob/b4a54a58696c72999a66ffec6fb8020a8a35c7c7/src/TestSessionEnvironment.php#L358

dnsl48 commented on src/TestSessionEnvironment.php:
```diff
@@ -252,26 +258,14 @@ protected function moveRecursive($src, $dest, $ignore = [])
+     * @return void
      */
-    public function applyState($state)
+    public function mergeWithExistingState($state)
```
Yeah, I guess we'll have to leave invocation for this method within the `applyState` for now to keep backward compatibility. Otherwise, it will be a breaking change, since `applyState` is a public method.

dnsl48 replied:
Do we need this as a public API here? Let's remove it?

dnsl48 commented on src/TestSessionEnvironment.php:
```diff
@@ -163,7 +166,9 @@ public function startTestSession($state = null, $id = null)
         $state = json_decode($json);
 
-        $this->applyState($state);
+        $state = $this->createDatabase($state);
```
I'd guess you also need `mergeWithExistingState` in here

blueo replied:
> I'd guess you also need `mergeWithExistingState` in here
hmm I'm not sure about that. mergeWithExistingState is mainly useful for adding several properties from another object - it doens't do anything with the state object other than mutate it, which is what createDatabase does. The request ends just after this so it is **maybe** not a problem but perhaps we need to have a 'setState' to persist these changes on the class property.

dnsl48 replied:
the only reason for that is backwards compatibility. No need for that anymore

dnsl48 commented on src/TestSessionEnvironment.php:
```diff
@@ -282,54 +276,93 @@ public function applyState($state)
+     * @throws InvalidArgumentException
+     * @return mixed $state
+     */
+    public function applyState($state)
```
Unfortunately, `applyState` is a public method and we cannot change its behaviour without introducing breaking changes.
What do you say if we rename this method to `setState` and then implement the `applyState` original behaviour just running the new API internally?
 - `createDatabase`
 - `mergeWithExistingState`
 - `setState`

In that case `applyState` behaviour would stay as is, so we keep backward compatibility and in the same time we introduce new API (and use it in start test session session instead of apply state?)

blueo replied:
I feel like backwards compatibility with this will be quite challenging. The current state has quite a lot of mixed logic. Create test session sets some things but then relies on the middleware and apply state to fill in the rest. I'd be more in favour of a major version bump as it seems a lot cleaner (and less work to re-implement the existing behaviour).

dnsl48 replied:
Yeah, I just discussed with Ingo. I guess it's gonna be much less resistant and more productive to issue a new major version when we need it. Let's rebase this on master and break stuff! :)

blueo replied:
rebased - should I target master with the PR?

dnsl48 replied:
yes please

blueo commented on src/TestSessionEnvironment.php:
```diff
@@ -252,26 +258,14 @@ protected function moveRecursive($src, $dest, $ignore = [])
-     * You can extend this by creating an Extension object and implementing either onBeforeApplyState() or
-     * onAfterApplyState() to add your own test state handling in.
+     * takes a state object (stdClass) and merges the existing state from getState()
+     * NB new state will not be overwritten
```
ah probably a bit of idiom: https://en.wikipedia.org/wiki/Nota_bene can take it out?

ScopeyNZ replied:
We've had a chat about this in our team recently (in the context of `iff`) - I think it's just better to write it in full 👍

dnsl48 commented on src/TestSessionEnvironment.php:
```diff
@@ -251,26 +257,14 @@ protected function moveRecursive($src, $dest, $ignore = [])
+     * @return void
      */
-    public function applyState($state)
+    public function mergeWithExistingState($state)
```
Do we need this as a public API here? Let's remove it?

dnsl48 commented:
Here's the fix for TestMailer - https://github.com/silverstripe/silverstripe-behat-extension/pull/188
It should be merged after this PR gets merged.

dnsl48 commented on src/TestSessionEnvironment.php:
```diff
@@ -282,43 +276,80 @@ public function applyState($state)
+        if (!$dbExists && $dbCreate) {
+            $this->oldDatabaseName = $databaseConfig['database'];
+            // Create a new one with a randomized name
+            $tempDB = new TempDatabase();
```
bump

blueo replied:
? This is changed... https://github.com/silverstripe/silverstripe-testsession/pull/65/files#diff-aa4d3caebbac216fecbec067ea5bd3daR314 @dnsl48

dnsl48 replied:
haha, that's another one :smile: 

https://github.com/silverstripe/silverstripe-testsession/pull/65/files#diff-7c0b5a4abfcf01ff6b876a5c12b7ed32R281

blueo replied:
oh so it is 🤦‍♂ fixed :)

dnsl48 commented:
Good to merge when https://github.com/silverstripe/silverstripe-behat-extension/pull/188 is merged

blueo commented:
@dnsl48 it feels like we might be able to merge/tag this before silverstripe/silverstripe-behat-extension#188 ? I'd be interested in doing a little more work on this module around assets etc but I'm worried this will become the never-ending-pr if I do...

dnsl48 commented:
> feels like we might be able to merge/tag this before silverstripe/silverstripe-behat-extension#188 ?

Sorry, no can do, without 188 it breaks email related tests. On the other hand, I'm quite certain this PR will be merged "as is", so if you base your other PRs on top of this one, you most likely won't even need to rebase

blueo commented:
but if this is a new major version - you don't need to use it until you want right? So the email related tests would still use the older version?

dnsl48 commented:
That's right. I think you're right. We can deal with behat-extension separately.

blueo commented:
cool thanks!
